unixfs/io/sizeadj.go

```diff
@@ -0,0 +1,112 @@
+package io
+
+import (
+	"errors"
+	"io"
+)
+
+func newSizeAdjReadSeekCloser(base ReadSeekCloser, size uint64) *sizeAdjReadSeekCloser {
+	r := &sizeAdjReadSeekCloser{base: base, size: int64(size)}
+	if r.size < 0 {
+		panic("size limited 2^63 − 1")
+	}
+	return r
+}
+
+type sizeAdjReadSeekCloser struct {
+	base   ReadSeekCloser
+	size   int64
+	offset int64
+}
+
+// Read implements the Read method as defined by io.Reader
+func (r *sizeAdjReadSeekCloser) Read(p []byte) (int, error) {
+	if r.offset >= r.size { // EOF
+		return 0, io.EOF
+	}
+	if int64(len(p)) > r.size-r.offset { // truncate
+		newsize := r.size - r.offset
+		p = p[:newsize]
+	}
+	n, err := r.base.Read(p)
+	if err != io.EOF { // only pad when we get an EOF
+		r.offset += int64(n)
+		return n, err
+	}
+	for ; n < len(p) && r.offset+int64(n) < r.size; n++ { // pad
+		p[n] = 0
+	}
+	r.offset += int64(n)
+	return n, io.EOF
+}
+
+// Seek implements the Seek method as defined by io.Seeker
+func (r *sizeAdjReadSeekCloser) Seek(offset int64, whence int) (int64, error) {
+	newOffset := r.offset
+	switch whence {
+	case io.SeekStart:
+		newOffset = offset
+	case io.SeekCurrent:
+		newOffset += offset
+	case io.SeekEnd:
+		newOffset = r.size + offset
+	}
+	if newOffset < 0 {
+		return -1, errors.New("Seek will result in negative position")
+	}
+	// Its easier just to always use io.SeekStart rather than
+	// correctly adjust offset for io.SeekCurrent and io.SeekEnd.
+	_, err := r.base.Seek(newOffset, io.SeekStart)
+	if err != nil {
+		return -1, err
+	}
+	r.offset = newOffset
+	return r.offset, nil
+}
+
+// Close implements the Close method as defined by io.Closer
+func (r *sizeAdjReadSeekCloser) Close() error {
+	return r.base.Close()
+}
+
+// WriteTo implemented WriteTo method as defined by io.WriterTo
+func (r *sizeAdjReadSeekCloser) WriteTo(w io.Writer) (int64, error) {
+	lr := &truncWriter{base: w, size: r.size - r.offset}
+	_, err := r.base.WriteTo(lr)
+	n := lr.offset
+	if err != nil {
+		r.offset += n
+		return n, err
+	}
+	if r.offset+n < r.size {
+		zeros := make([]byte, r.size-(r.offset+n))
```

Comment:
I am not sure what that limit is, and I don't see a strong need for it, however, I can be convinced otherwise.

Reply:
what happens here if `r.size == 10000000000` and `r.base.Size() == 1` and we end up here? smells like OOMdeath

Reply:
this could be trivially used to attack a gateway (a user could construct such a malicious node, and then just try to fetch it through the gateway. The gateway uses `WriteTo` as an optimization)

Reply:
Acknowledge.  I am just not sure what the solution should be.

Reply:
just something that writes a bunch of zeros without allocating unboundedly? like maybe a `WriteZeros` method that takes a write and a count, and writes zeros in 1k chunks up to the desired amount

Reply:
Okay done.